node/src/Pos/Block/Network/Retrieval.hs

```diff
@@ -89,75 +89,111 @@ retrievalWorkerImpl SendActions {..} =
         queue        <- view (lensOf @BlockRetrievalQueueTag)
         recHeaderVar <- view (lensOf @RecoveryHeaderTag)
         logDebug "Waiting on the block queue or recovery header var"
+        -- Reading the queue is a priority, because it sets recovery
+        -- variable in case header is alternative. So if the queue
+        -- contains lots of headers after a long delay, we'll first
+        -- iterate over them and set recovery variable to a latest
+        -- one, and only then we'll do recovery.
         thingToDoNext <- atomically $ do
             mbQueuedHeadersChunk <- tryReadTBQueue queue
             mbRecHeader <- tryReadTMVar recHeaderVar
             case (mbQueuedHeadersChunk, mbRecHeader) of
                 (Nothing, Nothing) -> retry
+                -- Throw away continuation request, if recovery var is
+                -- not empty. By definition, recovery header is newer
+                -- than continuation so we prefer it.
+                (Just (_, BlockRetrievalTask{..}), Just _) | brtContinues -> retry
+                -- Dispatch task.
                 (Just (nodeId, task), _) ->
-                    pure (handleBlockRetrievalFromQueue nodeId task)
+                    pure (handleBlockRetrieval nodeId task)
+                -- No tasks & recovery header is there ⇒ do recovery.
                 (_, Just (nodeId, rHeader))  ->
-                    pure (handleHeadersRecovery nodeId rHeader)
+                    pure (handleRecoveryWithHandler nodeId rHeader)
         thingToDoNext
         mainLoop
     mainLoopE e = do
         -- REPORT:ERROR 'reportOrLogE' in block retrieval worker.
-        reportOrLogE "retrievalWorker: error caught " e
-        delay (30 & sec)
+        reportOrLogE "retrievalWorker mainLoopE: error caught " e
+        delay $ sec 1
         mainLoop
-    --
-    handleBlockRetrieval nodeId BlockRetrievalTask{..} =
-        handleAll (handleBlockRetrievalE nodeId brtHeader) $
+
+    -----------------
+
+    -- That's the first queue branch (task dispatching).
+    handleBlockRetrieval nodeId BlockRetrievalTask{..} = do
```

Comment:
Can you explain why we've dropped the `handleAll (handleBlockRetrievalE ...` hander.

Reply:
Because we don't need it in fact. 
Previously this handler was handling both branches (task dispatching and recovery), now it's renamed into `handleRecoveryE` and is used for recovery only. What this handler was doing is, in fact, trying to restart recovery. 
Now errors from task dispatching are catched directly in `mainLoopE`.